// api-dylux/main.js
const ds = require('./index.js')
const tiktok = require("./lib/tiktok")
const emojimix = require("./lib/emojimix")


module.exports.xvideosSearch = ds.xvideosSe
module.exports.xnxxSearch = ds.xnxxSe
module.exports.xvideosdl = ds.xvideosdl
module.exports.xnxxdl = ds.xnxxdl
module.exports.emojimix = emojimix  //
module.exports.tiktok = tiktok
module.exports.GDriveDl = ds.Gdrive
module.exports.yta = ds.ytMp3
module.exports.ytv = ds.ytMp4
module.exports.fbdl = ds.facebook
module.exports.twitter = ds.twitter
module.exports.SoundCloud = ds.Downloader
module.exports.Imgur = ds.Downloader
module.exports.Imdb = ds.Downloader
module.exports.Anime = ds.Anime
module.exports.Manga = ds.Manga
module.exports.Character = ds.Character
module.exports.JadwalBola = ds.JadwalBola
module.exports.JadwalTv = ds.JadwalTv
module.exports.JadwalSholat = ds.JadwalSholat
module.exports.Pinterest = ds.Pinterest
module.exports.Film = ds.Film
module.exports.Wattpad = ds.Wattpad
module.exports.WattpadUser = ds.WattpadUser
module.exports.Webtoons = ds.Webtoons
module.exports.Mangatoons = ds.Mangatoons
module.exports.Drakor = ds.Drakor
module.exports.Telesticker = ds.telesticker
module.exports.StickerSearch = ds.StickerSearch
module.exports.ListSurah = ds.ListSurah
module.exports.Surah = ds.Surah
module.exports.TafsirSurah = ds.TafsirSurah
module.exports.KompasNews = ds.KompasNews
module.exports.Covid = ds.Covid
module.exports.Shoope = ds.Shoope
module.exports.Gempa = ds.Gempa
module.exports.GSMArena = ds.GSMArena
module.exports.webp2mp4 = ds.webp2mp4
module.exports.webp2png = ds.webp2png
module.exports.Emoji = ds.Emoji
module.exports.Wikipedia = ds.Wikipedia
module.exports.RandomCerpen = ds.RandomCerpen
module.exports.Mediafire = ds.mediafireDl
module.exports.Mangatoon = ds.MangaToon
module.exports.igStalk = ds.igStalk
module.exports.ttStalk = ds.tiktokStalk
module.exports.sfileDl = ds.sfileDl
module.exports.sfileSearch = ds.sfileSe
module.exports.ResepSearch = ds.ResepSearch
module.exports.ResepRead = ds.ReadResep
module.exports.SearchGore = ds.SearchGore
module.exports.Goredl = ds.Goredl
module.exports.RandomGore = ds.RandomGore
module.exports.GrupWA = ds.GrupWA
module.exports.ArtiNama = ds.ArtiNama
module.exports.WallpaperHd = ds.wallpaperhd
module.exports.Tt2 = ds.lovetik
module.exports.igdl = ds.igdl
module.exports.Kbbi = ds.kbbi
module.exports.ttp = ds.ttp
