// @bochilteam/scraper/lib/cjs/games/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.tebaktebakanjson = exports.tebaktebakan = exports.tebaklirikjson = exports.tebaklirik = exports.tebakbenderajson = exports.tebakbendera = exports.susunkatajson = exports.susunkata = exports.siapakahakujson = exports.siapakahaku = exports.tebakkabupatenjson = exports.tebakkabupaten = exports.tebakkimiajson = exports.tebakkimia = exports.tekatekijson = exports.tekateki = exports.tebakkatajson = exports.tebakkata = exports.asahotakjson = exports.asahotak = exports.family100json = exports.family100 = exports.caklontongjson = exports.caklontong = exports.tebakgambarjson = exports.tebakgambar = void 0;
const asahotak_js_1 = __importStar(require("./asahotak.js"));
exports.asahotak = asahotak_js_1.default;
Object.defineProperty(exports, "asahotakjson", { enumerable: true, get: function () { return asahotak_js_1.asahotakjson; } });
const caklontong_js_1 = __importStar(require("./caklontong.js"));
exports.caklontong = caklontong_js_1.default;
Object.defineProperty(exports, "caklontongjson", { enumerable: true, get: function () { return caklontong_js_1.caklontongjson; } });
const family100_js_1 = __importStar(require("./family100.js"));
exports.family100 = family100_js_1.default;
Object.defineProperty(exports, "family100json", { enumerable: true, get: function () { return family100_js_1.family100json; } });
const siapakahaku_js_1 = __importStar(require("./siapakahaku.js"));
exports.siapakahaku = siapakahaku_js_1.default;
Object.defineProperty(exports, "siapakahakujson", { enumerable: true, get: function () { return siapakahaku_js_1.siapakahakujson; } });
const susunkata_js_1 = __importStar(require("./susunkata.js"));
exports.susunkata = susunkata_js_1.default;
Object.defineProperty(exports, "susunkatajson", { enumerable: true, get: function () { return susunkata_js_1.susunkatajson; } });
const tebakbendera_js_1 = __importStar(require("./tebakbendera.js"));
exports.tebakbendera = tebakbendera_js_1.default;
Object.defineProperty(exports, "tebakbenderajson", { enumerable: true, get: function () { return tebakbendera_js_1.tebakbenderajson; } });
const tebakgambar_js_1 = __importStar(require("./tebakgambar.js"));
exports.tebakgambar = tebakgambar_js_1.default;
Object.defineProperty(exports, "tebakgambarjson", { enumerable: true, get: function () { return tebakgambar_js_1.tebakgambarjson; } });
const tebakkabupaten_js_1 = __importStar(require("./tebakkabupaten.js"));
exports.tebakkabupaten = tebakkabupaten_js_1.default;
Object.defineProperty(exports, "tebakkabupatenjson", { enumerable: true, get: function () { return tebakkabupaten_js_1.tebakkabupatenjson; } });
const tebakkata_js_1 = __importStar(require("./tebakkata.js"));
exports.tebakkata = tebakkata_js_1.default;
Object.defineProperty(exports, "tebakkatajson", { enumerable: true, get: function () { return tebakkata_js_1.tebakkatajson; } });
const tebakkimia_js_1 = __importStar(require("./tebakkimia.js"));
exports.tebakkimia = tebakkimia_js_1.default;
Object.defineProperty(exports, "tebakkimiajson", { enumerable: true, get: function () { return tebakkimia_js_1.tebakkimiajson; } });
const tebaklirik_js_1 = __importStar(require("./tebaklirik.js"));
exports.tebaklirik = tebaklirik_js_1.default;
Object.defineProperty(exports, "tebaklirikjson", { enumerable: true, get: function () { return tebaklirik_js_1.tebaklirikjson; } });
const tebaktebakan_js_1 = __importStar(require("./tebaktebakan.js"));
exports.tebaktebakan = tebaktebakan_js_1.default;
Object.defineProperty(exports, "tebaktebakanjson", { enumerable: true, get: function () { return tebaktebakan_js_1.tebaktebakanjson; } });
const tekateki_js_1 = __importStar(require("./tekateki.js"));
exports.tekateki = tekateki_js_1.default;
Object.defineProperty(exports, "tekatekijson", { enumerable: true, get: function () { return tekateki_js_1.tekatekijson; } });
