// xfarr-api/main.js
const xa = require('./index.js')

module.exports.Youtube = xa.Downloader
module.exports.Tiktok = xa.Downloader
module.exports.Instagram = xa.Downloader
module.exports.Facebook = xa.Downloader
module.exports.Twitter = xa.Downloader
module.exports.SoundCloud = xa.Downloader
module.exports.Imgur = xa.Downloader
module.exports.Imdb = xa.Downloader
module.exports.Anime = xa.Anime
module.exports.Manga = xa.Manga
module.exports.Character = xa.Character
module.exports.JadwalBola = xa.JadwalBola
module.exports.JadwalTv = xa.JadwalTv
module.exports.JadwalSholat = xa.JadwalSholat
module.exports.Pinterest = xa.Pinterest
module.exports.Film = xa.Film
module.exports.Wattpad = xa.Wattpad
module.exports.WattpadUser = xa.WattpadUser
module.exports.Webtoons = xa.Webtoons
module.exports.Mangatoons = xa.Mangatoons
module.exports.Drakor = xa.Drakor
module.exports.Telesticker = xa.Telesticker
module.exports.StickerSearch = xa.StickerSearch
module.exports.ListSurah = xa.ListSurah
module.exports.Surah = xa.Surah
module.exports.TafsirSurah = xa.TafsirSurah
module.exports.KompasNews = xa.KompasNews
module.exports.INews = xa.INews
