// dandi-api/main.js
const ds = require('./index.js')


module.exports.Yta = ds.ytdown
module.exports.Ytv = ds.ytdown
module.exports.Yta2 = ds.yta
module.exports.Ytv2 = ds.ytv
module.exports.Facebook = ds.Facebook
module.exports.Twitter = ds.Downloader
module.exports.SoundCloud = ds.Downloader
module.exports.Imgur = ds.Downloader
module.exports.Imdb = ds.Downloader
module.exports.Anime = ds.Anime
module.exports.Manga = ds.Manga
module.exports.Character = ds.Character
module.exports.JadwalBola = ds.JadwalBola
module.exports.JadwalTv = ds.JadwalTv
module.exports.JadwalSholat = ds.JadwalSholat
module.exports.Pinterest = ds.Pinterest
module.exports.Film = ds.Film
module.exports.Wattpad = ds.Wattpad
module.exports.WattpadUser = ds.WattpadUser
module.exports.Webtoons = ds.Webtoons
module.exports.Mangatoons = ds.Mangatoons
module.exports.Drakor = ds.Drakor
module.exports.Telesticker = ds.telesticker
module.exports.StickerSearch = ds.StickerSearch
module.exports.ListSurah = ds.ListSurah
module.exports.Surah = ds.Surah
module.exports.TafsirSurah = ds.TafsirSurah
module.exports.KompasNews = ds.KompasNews
module.exports.Covid = ds.Covid
module.exports.Shoope = ds.Shoope
module.exports.Gempa = ds.Gempa
module.exports.GSMArena = ds.GSMArena
module.exports.webp2mp4 = ds.webp2mp4
module.exports.webp2png = ds.webp2png
module.exports.Emoji = ds.Emoji
module.exports.Wikipedia = ds.Wikipedia
module.exports.RandomCerpen = ds.RandomCerpen
module.exports.Mediafire = ds.mediafireDl
module.exports.Mangatoon = ds.MangaToon
module.exports.IgStalk = ds.IgStalk
module.exports.SfileDown = ds.SfileDown
module.exports.SfileSearch = ds.SfileSearch
module.exports.ResepSearch = ds.ResepSearch
module.exports.ResepRead = ds.ReadResep
module.exports.SearchGore = ds.SearchGore
module.exports.Goredl = ds.Goredl
module.exports.RandomGore = ds.RandomGore
module.exports.GrupWA = ds.GrupWA
module.exports.ArtiNama = ds.ArtiNama
module.exports.WallpaperHd = ds.wallpaperhd
module.exports.Tiktok = ds.TiktokDl
module.exports.Tt2 = ds.lovetik
module.exports.Instagram = ds.igdl
module.exports.Kbbi = ds.kbbi
