// @bochilteam/scraper/lib/cjs/others/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.zippyshare = exports.nameFreeFire = exports.kbbi = exports.listJadwalTV = exports.jadwalTVNow = exports.jadwalTV = exports.wikipedia = void 0;
var wikipedia_js_1 = require("./wikipedia.js");
Object.defineProperty(exports, "wikipedia", { enumerable: true, get: function () { return __importDefault(wikipedia_js_1).default; } });
var jadwal_tv_js_1 = require("./jadwal-tv.js");
Object.defineProperty(exports, "jadwalTV", { enumerable: true, get: function () { return __importDefault(jadwal_tv_js_1).default; } });
Object.defineProperty(exports, "jadwalTVNow", { enumerable: true, get: function () { return jadwal_tv_js_1.jadwalTVNow; } });
Object.defineProperty(exports, "listJadwalTV", { enumerable: true, get: function () { return jadwal_tv_js_1.listJadwalTV; } });
var KBBI_js_1 = require("./KBBI.js");
Object.defineProperty(exports, "kbbi", { enumerable: true, get: function () { return __importDefault(KBBI_js_1).default; } });
var idff_js_1 = require("./idff.js");
Object.defineProperty(exports, "nameFreeFire", { enumerable: true, get: function () { return __importDefault(idff_js_1).default; } });
var zippyshare_js_1 = require("./zippyshare.js");
Object.defineProperty(exports, "zippyshare", { enumerable: true, get: function () { return __importDefault(zippyshare_js_1).default; } });
__exportStar(require("./minecraft.js"), exports);
__exportStar(require("./mediafire.js"), exports);
__exportStar(require("./BMKG.js"), exports);
__exportStar(require("./lyrics.js"), exports);
__exportStar(require("./minecraft.js"), exports);
__exportStar(require("./bioskop.js"), exports);
__exportStar(require("./chord.js"), exports);
// TODO: add cekResi
// export * from "./cek-resi"
