// insta-fetcher/dist/utils/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseCookie = exports.formatCookie = exports.bufferToStream = exports.randInt = exports.getPostType = exports.isIgPostUrl = exports.shortcodeFormatter = exports.IGPostRegex = exports.shortcodeFromMediaID = exports.shortcodeToMediaID = void 0;
const stream_1 = require("stream");
const index_1 = require("../types/index");
const big_integer_1 = __importDefault(require("big-integer"));
// https://stackoverflow.com/questions/16758316/where-do-i-find-the-instagram-media-id-of-a-image
// https://gist.github.com/sclark39/9daf13eea9c0b381667b61e3d2e7bc11
const lower = 'abcdefghijklmnopqrstuvwxyz';
const upper = lower.toUpperCase();
const numbers = '0123456789';
const ig_alphabet = upper + lower + numbers + '-_';
const bigint_alphabet = numbers + lower;
/**
 * convert instagram shortcode into media_id
 * @param shortcode
 * @returns
 */
const shortcodeToMediaID = (shortcode) => {
    const o = shortcode.replace(/\S/g, m => {
        var c = ig_alphabet.indexOf(m);
        var b = bigint_alphabet.charAt(c);
        return (b != "") ? b : `<${c}>`;
    });
    return (0, big_integer_1.default)(o, 64).toString(10);
};
exports.shortcodeToMediaID = shortcodeToMediaID;
const shortcodeFromMediaID = (media_id) => {
    var o = (0, big_integer_1.default)(media_id).toString(64);
    return o.replace(/<(\d+)>|(\w)/g, (_m, m1, m2) => {
        return ig_alphabet.charAt((m1)
            ? parseInt(m1)
            : bigint_alphabet.indexOf(m2));
    });
};
exports.shortcodeFromMediaID = shortcodeFromMediaID;
/** Instagram post regex */
exports.IGPostRegex = /(?:https?:\/\/)?(?:www\.)?(?:instagram\.com(?:\/.+?)?\/(p|reel|tv)\/)([\w-]+)(?:\/)?(\?.*)?$/gim;
/**
 * format instagram long url to get shortcode
 * @param url a instagram post url
 * @returns {formattedShortcode}
 */
const shortcodeFormatter = (url) => {
    const re = /(?:https?:\/\/)?(?:www\.)?(?:instagram\.com(?:\/.+?)?\/(p|reel|tv)\/)([\w-]+)(?:\/)?(\?.*)?$/gim.exec(url) || '';
    return {
        type: re[1],
        shortcode: re[2],
        url: 'https://www.instagram.com/' + re[1] + '/' + re[2],
        media_id: (0, exports.shortcodeToMediaID)(re[2])
    };
};
exports.shortcodeFormatter = shortcodeFormatter;
/**
 * is Instagram Url?
 * @param url instagram post url
 * @returns
 */
const isIgPostUrl = (url) => {
    return /(?:https?:\/\/)?(?:www\.)?(?:instagram\.com(?:\/.+?)?\/(p|reel|tv)\/)([\w-]+)(?:\/)?(\?.*)?$/gim.test(url);
};
exports.isIgPostUrl = isIgPostUrl;
/**
 * get instagram post type
 * @param type product_type
 * @returns
 */
const getPostType = (type) => {
    return type == index_1.ProductType.CAROUSEL
        ? index_1.IGPostType.carousel_container
        : type == index_1.ProductType.REEL
            ? index_1.IGPostType.clips
            : type == index_1.ProductType.SINGLE
                ? index_1.IGPostType.feed
                : type == index_1.ProductType.TV
                    ? index_1.IGPostType.igtv
                    : index_1.IGPostType.feed;
};
exports.getPostType = getPostType;
/** get random number in range */
const randInt = (min, max, q = 0.001) => {
    return Math.floor((Math.random() * (min - max)) / q) * q;
};
exports.randInt = randInt;
// https://stackoverflow.com/questions/13230487/converting-a-buffer-into-a-readablestream-in-node-js
const bufferToStream = (buffer) => {
    const readable = new stream_1.Readable();
    readable._read = () => { }; // _read is required but you can noop it
    readable.push(buffer);
    readable.push(null);
    return readable;
};
exports.bufferToStream = bufferToStream;
const formatCookie = (setCookie) => {
    var _a;
    return (_a = setCookie === null || setCookie === void 0 ? void 0 : setCookie.map(x => { var _a; return (_a = x.match(/(.*?=.*?);/)) === null || _a === void 0 ? void 0 : _a[1]; })) === null || _a === void 0 ? void 0 : _a.join('; ');
};
exports.formatCookie = formatCookie;
const parseCookie = (str) => {
    return str.split(';')
        .map(v => v.trim().split('='))
        .reduce((acc, v) => {
        acc[decodeURIComponent(v[0])] = decodeURIComponent(v[1]);
        delete acc[''];
        return acc;
    }, {});
};
exports.parseCookie = parseCookie;
